### DISCUSSION TITLE
$model->role() doesn't work within job or invokable function

### DISCUSSION THREAD

**Discussion by ankheur** [spatie/laravel-permission#2472] [Q&A]
Hi everyone !

I'm using a cron task to dispatch a job everyday to send emails to people depending on a condition. 
Laravel 9.51
Spatie Laravel-permission 5.9.1

Here is a sample of the code used:

```php
$events = Events::where('date_beginning', '>', \Carbon\Carbon::now())->with('companies')->get();
foreach($events as $event) {
	$companies = $event->companies()->whereHas('invite_requests', function ($query) {
                   $query->where('status', '=', 'pending');
        })->with('contacts')->get();

        foreach($companies as $company) {
             $contact = $company->contacts()->role('referer')->first();
             if($contact) {
                 Log::info('Email sent to  ' . $company->name. ' - ' . $contact->email);
                 Mail::to($contact->email)->queue(new InviteRequestEmail($contact));
             }
             else {
                 Log::error('No referer for ' . $company->name);
             }
       }			
}
```

But within the job or even from an invokable function, this code doesn't work and I've narrowed the problem down to the :
`$company->contacts()->role('referer')->first();`
It seems that the job cannot access the roles but this same code works perfectly when I put it in a controller method and call it from a url !

What could be the problem? Thanks a lot for your help!

**Comment by parallels999** [comment]
Are you using teams? version?

**Reply by ankheur** [reply]
I'm using :
Laravel 9.51
Spatie Laravel-permission 5.9.1

And not using teams

**Reply by parallels999** [reply]
I'm thinking that it never enters that `for`

**Comment by erikn69** [comment]
> I'm using a cron task to dispatch a job

The task has his own cache instance, try clearing cache on every run, try
```php
app()->make(\Spatie\Permission\PermissionRegistrar::class)->forgetCachedPermissions();
// or
app()->make(\Spatie\Permission\PermissionRegistrar::class)->clearClassPermissions();
```

**Reply by ankheur** [reply]
It unfortunately doesn't make it work. The problem lies on the role that the job doesn't seem to be able to access, not even once and just stop the script. It doesn't even return an empty response, it just stops.

**Reply by drbyte** [reply]
If the script "just stops", then it's probably triggering a fatal PHP error, which should be in your logs. What's the error?

**Reply by ankheur** [reply]
That's the problem, there is no error thrown ! Nothing in the laravel logs nor the server logs. I've tried encapsulating the ```$contact = $company->contacts()->role('referer')->first();``` line with a try catch but nothing

**Reply by erikn69** [reply]
Try `whereHas` directly
```php
$company->contacts()
    ->whereHas('roles', fn ($query) => $query
        ->where('name', 'referer')->where('guard_name', 'YOUR_GUARD_NAME')
    )->first()
```

**Reply by ankheur** [reply]
Thanks for your reply ! It unfortunately doesn't work either

**Comment by drbyte** [comment]
If `$company->contacts()->role('referer')->first();` is failing because of the call to the `role` scope, then perhaps some debugging is needed to see what sort of object you have. What shows when you dump `$company->contacts()` to the log? How about dumping `$company->contacts()->with('roles')->get()` ?

**Reply by ankheur** [reply]
```$company->contacts()``` gives me what the companies I expect in a collection with all of their contacts, as intended.
But the ```$company->contacts()->with('roles')->get()``` doesn't work. Same as before, it stops, no error, nothing in the logs.

**Reply by drbyte** [reply]
how about `$company->contacts()->first()->roles;` ?

**Reply by drbyte** [reply]
... and if that just stops, what about just `$company->contacts()->first()` .... and then inspect that object to see whether it has the necessary things that roles depend on (Authenticatable object, HasRoles trait, etc)

**Reply by ankheur** [reply]
The thing is that all of that works when I put the code in a regular controller method and display it via a url. 
But when I use it via a job or an invokable class called from the scheduler, then it doesn't anymore.

**Reply by drbyte** [reply]
Yes, that's the hard part about Jobs: they lose state/context compared to an application hit that comes thru an http request.

That's why I'm trying asking questions to reverse-engineer where some needed contextual part is missing.

**Reply by ankheur** [reply]
And thanks a lot for helping me !

To answer your questions, both don't work but ```$exposant->contacts->first()``` does and gives me the first contact of the company

**Reply by drbyte** [reply]
Do you mind posting the whole job class? (I can comprehend some french, so probably don't need to translate.)
And maybe also the Contact class?

**Reply by ankheur** [reply]
The job class 
```php
namespace App\Jobs;

use App\Mail\RappelDemandesInvitation;
use App\Models\Manifestation;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Mail;
use Illuminate\Support\Facades\Log;

class MailingRappelDemandesInvitation implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    /**
     * Create a new job instance.
     *
     * @return void
     */
    public function __construct()
    {
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        $manifestations = Manifestation::where('date_debut', '>', \Carbon\Carbon::now())->with('exposants')->get();
        Log::info('Job demandes invitation lancé');

        foreach($manifestations as $manifestation) {
            Log::info('Envoie mail demandes invitation pour '.$manifestation->slug);

            $exposants = $manifestation->exposants()->whereHas('demandes_invitation', function ($query) {
                $query->where('statut', '=', 'en attente');
            })->with('contacts')->get();

            foreach ($exposants as $exposant) {
                $contact = $exposant->contacts()->role('contact_logistique')->first();

                if ($contact) {
                    Log::info('Envoie mail demandes invitation pour '.$exposant->denomination.'(#'.$exposant->id.') - '.$contact->email);
                    Mail::to($contact->email)->send(new RappelDemandesInvitation($contact));
                } else {
                    Log::error('Pas de contact logistique pour '.$exposant->denomination.' (#'.$exposant->id.')');
                }
            }
        }
    }
}
```

And the contact class
```php
namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Notifications\Notifiable;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Support\Facades\Hash;
use Spatie\Permission\Traits\HasRoles;

class ExposantContact extends Authenticatable
{
    use HasFactory, HasRoles, SoftDeletes, Notifiable;

    protected $table = 'exposant_contacts';

    protected $guard = 'exposant';

    protected $guarded = [];

    protected $hidden = [
        'password', 'remember_token',
    ];

    public function setPasswordAttribute($value)
    {
        $this->attributes['password'] = Hash::make($value);
    }

    public function exposant()
    {
        return $this->belongsTo(Exposant::class);
    }
}
```

**Reply by drbyte** [reply]
1. Does it make any difference if you remove `implements ShouldQueue` from the job? That should force the job to run immediately instead of running via a queue-worker ... which might affect "state".
2. What's the `guard_name` on the role record for the role named `contact_logistique`?
3. Does it make any difference (for both web and for jobs) if you change the `ExposantContact` class's property `$guard` to `$guard_name` ?

**Reply by ankheur** [reply]
1) I've tried removing the ShouldQueue, not working :(
2) The guard name for the role is defined as 'exposant' in the roles table
3) Tried it, no difference for the job, still not running the role query

**Reply by drbyte** [reply]
Try:
```diff
-                $contact = $exposant->contacts()->role('contact_logistique')->first();
+                $contact = $exposant->contacts()->role('contact_logistique')->ddRawSql();
```
What's the query that it outputs? (ddRawSql() exists since Laravel 10.15.0; there are other ways to output the query for older Laravel versions).
I'm curious what the actual query is. 
Perhaps try also dropping the `->role(...)` part to see how the query changes.
